gui/src/2D/controls/toggleButton.ts

```diff
@@ -0,0 +1,308 @@
+     */
+    public toInactiveAnimation: () => void;
+
+    /**
+     * Function called to generate a pointer enter animation when the toggle is active.
+     */
+    public pointerEnterActiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer out animation when the toggle is active.
+     */
+    public pointerOutActiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer down animation when the toggle is active.
+     */
+    public pointerDownActiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer up animation when the toggle is active.
+     */
+    public pointerUpActiveAnimation: () => void;
+
+    /**
+     * Function called to generate a pointer enter animation when the toggle is inactive.
+     */
+    public pointerEnterInactiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer out animation when the toggle is inactive.
+     */
+    public pointerOutInactiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer down animation when the toggle is inactive.
+     */
+    public pointerDownInactiveAnimation: () => void;
+    /**
+     * Function called to generate a pointer up animation when the toggle is inactive.
+     */
+    public pointerUpInactiveAnimation: () => void;
+
+    /** Observable raised when isActive is changed */
+    public onIsActiveChangedObservable = new Observable<boolean>();
+
+    /**
+     * Gets or sets a boolean indicating that the toggle will let internal controls handle picking instead of doing it directly using its bounding info
+     */
+    public delegatePickingToChildren = false;
+
+    private _image: Nullable<Image>;
+    /**
+     * Returns the image part of the button (if any)
+     */
+    public get image(): Nullable<Image> {
+        return this._image;
+    }
+
+    private _textBlock: Nullable<TextBlock>;
+    /**
+     * Returns the image part of the button (if any)
+     */
+    public get textBlock(): Nullable<TextBlock> {
+        return this._textBlock;
+    }
+
+    private _group: string;
+    /** Gets or sets group name this toggle belongs to */
+    public get group(): string {
+        return this._group;
+    }
+    public set group(value: string) {
+        if (this._group === value) {
+            return;
+        }
+
+        this._group = value;
+    }
+
+    private _isActive = false;
+    /** Gets or sets a boolean indicating if the toogle button is active or not */
+    public get isActive(): boolean {
+        return this._isActive;
+    }
+    public set isActive(value: boolean) {
+        // Function modeled after radioButton.ts
+        if (this._isActive === value) {
+            return;
+        }
+
+        this._isActive = value;
+
+        // Update the visual state based on the new value
+        if (this._isActive) {
+            this.toActiveAnimation();
+        } else {
+            this.toInactiveAnimation();
+        }
```

Comment:
You test event handlers for not being undefined further down, so I guess you should also do it here.

Reply:
Would an optional chaining operator be acceptable in the BabylonJS project? 

_The traditional way:_
`if (this._isActive) {
    if (this.toActiveAnimation) {
        this.toActiveAnimation();
    }
} else {
    if (this.toInactiveAnimation) {
        this.toInactiveAnimation()
    }
}`

_Using Optional Chaining:_
`if (this._isActive) {
    this.toActiveAnimation.?();
} else {
    this.toInactiveAnimation.?();
}`

From the [Mozilla's documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), it looks safe to use, but wasn't sure if there were compatibility concerns.

Reply:
should be good as transpiled by ts anyway

Reply:
This should be resolved in the latest update to this pull request.